# Project 2: Data Cleaning

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# read csvs
train = pd.read_csv('./datasets/train.csv')
test = pd.read_csv('./datasets/test.csv')

In [3]:
# set option to display max columns always
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

## Handling of train and test dataframe

For ease of handling, I will be merging the test and train dataframes into 1 concatenated dataframe. I split the dataset back into two in the Feature Engineering notebook after all feature engineering is done, to ensure that there are same columns for both dataframes.

In [4]:
train.shape

(2051, 81)

In [5]:
test.shape

(878, 80)

In [6]:
concat_df = pd.concat([train, test])

Following from the EDA notebook, lots of cleaning is required. The following are the cleaning steps that I would be performing, in no particular order.

1) Adding other relevant features

2) Handling null values

3) Handling outliers

4) Dropping columns

5) Saving data


### Additional features for age of property

In [7]:
# Adding structure age variable depending if there was a major remodeling
concat_df['Age'] = concat_df.apply(lambda x: x['Yr Sold']-x['Year Built'] if (x['Year Built']<x['Year Remod/Add']) 
                                                           else (x['Yr Sold']-x['Year Remod/Add']), axis=1)

### Dropping features after EDA

The following columns will be dropped as there is no clear relationship between sales price and the feature. As such, to prevent noise in the dataset, they will be dropped early in the cleaning phase. They are:

1) Id

2) PID


Year related columns will also be dropped as a processed 'Age' column essentially captures all necessary information already.

3) Yr Sold

4) Year Built

5) Year Remod/Add

In [8]:
concat_df = concat_df.drop(['Id',
                    'PID',
                    'Yr Sold',
                    'Year Built',
                    'Year Remod/Add'                            
                       ],
                       axis=1)

In [9]:
concat_df.shape

(2929, 77)

In [10]:
# Filter out numeric and categorical features column names
num_features = [col for col in concat_df._get_numeric_data().columns]
cat_features = [col for col in concat_df.select_dtypes(include=['object']).columns]

### Cleaning concatenated dataset

In [11]:
concat_df.shape

(2929, 77)

In [12]:
concat_df.describe().transpose()

count           mean           std      min       25%  \
MS SubClass      2929.0      57.386480     42.645278     20.0      20.0   
Lot Frontage     2439.0      69.199672     23.337677     21.0      58.0   
Lot Area         2929.0   10137.697849   7861.902406   1300.0    7440.0   
Overall Qual     2929.0       6.093547      1.409420      1.0       5.0   
Overall Cond     2929.0       5.563332      1.111678      1.0       5.0   
Mas Vnr Area     2906.0     101.669649    178.724156      0.0       0.0   
BsmtFin SF 1     2928.0     442.008880    454.428364      0.0       0.0   
BsmtFin SF 2     2928.0      49.739413    169.194874      0.0       0.0   
Bsmt Unf SF      2928.0     559.153689    439.529726      0.0     219.0   
Total Bsmt SF    2928.0    1050.901981    438.999166      0.0     793.0   
1st Flr SF       2929.0    1158.882212    390.248168    334.0     876.0   
2nd Flr SF       2929.0     335.045408    427.891935      0.0       0.0   
Low Qual Fin SF  2929.0       4.678389     46.318337      0.0       0.0   
Gr Liv Area      2929.0    1498.606009    502.174926    334.0    1126.0   
Bsmt Full Bath   2927.0       0.431158      0.524805      0.0       0.0   
Bsmt Half Bath   2927.0       0.061155      0.245293      0.0       0.0   
Full Bath        2929.0       1.566064      0.552400      0.0       1.0   
Half Bath        2929.0       0.379310      0.502584      0.0       0.0   
Bedroom AbvGr    2929.0       2.854216      0.827868      0.0       2.0   
Kitchen AbvGr    2929.0       1.044384      0.214111      0.0       1.0   
TotRms AbvGrd    2929.0       6.441448      1.570977      2.0       5.0   
Fireplaces       2929.0       0.599181      0.647989      0.0       0.0   
Garage Yr Blt    2770.0    1978.122022     25.527124   1895.0    1960.0   
Garage Cars      2928.0       1.766393      0.760355      0.0       1.0   
Garage Area      2928.0     472.679303    214.948914      0.0     320.0   
Wood Deck SF     2929.0      93.712871    126.365495      0.0       0.0   
Open Porch SF    2929.0      47.411062     67.168919      0.0       0.0   
Enclosed Porch   2929.0      23.019461     64.148601      0.0       0.0   
3Ssn Porch       2929.0       2.593377     25.145578      0.0       0.0   
Screen Porch     2929.0      16.007511     56.096167      0.0       0.0   
Pool Area        2929.0       2.244111     35.603235      0.0       0.0   
Misc Val         2929.0      50.652441    566.440219      0.0       0.0   
Mo Sold          2929.0       6.214749      2.714055      1.0       4.0   
SalePrice        2051.0  181469.701609  79258.659352  12789.0  129825.0   
Age              2929.0      36.446910     30.288718     -1.0       7.0   

                      50%        75%       max  
MS SubClass          50.0      70.00     190.0  
Lot Frontage         68.0      80.00     313.0  
Lot Area           9434.0   11553.00  215245.0  
Overall Qual          6.0       7.00      10.0  
Overall Cond          5.0       6.00       9.0  
Mas Vnr Area          0.0     164.00    1600.0  
BsmtFin SF 1        370.0     734.00    5644.0  
BsmtFin SF 2          0.0       0.00    1526.0  
Bsmt Unf SF         465.5     801.25    2336.0  
Total Bsmt SF       990.0    1300.50    6110.0  
1st Flr SF         1084.0    1384.00    5095.0  
2nd Flr SF            0.0     703.00    2065.0  
Low Qual Fin SF       0.0       0.00    1064.0  
Gr Liv Area        1442.0    1742.00    5642.0  
Bsmt Full Bath        0.0       1.00       3.0  
Bsmt Half Bath        0.0       0.00       2.0  
Full Bath             2.0       2.00       4.0  
Half Bath             0.0       1.00       2.0  
Bedroom AbvGr         3.0       3.00       8.0  
Kitchen AbvGr         1.0       1.00       3.0  
TotRms AbvGrd         6.0       7.00      15.0  
Fireplaces            1.0       1.00       4.0  
Garage Yr Blt      1979.0    2002.00    2207.0  
Garage Cars           2.0       2.00       5.0  
Garage Area         480.0     576.00    1488.0  
Wood Deck SF          0.0     168.00    142

There are simply too many zero values and missing/ null values as seen from count and min values in the data above. Further cleaning efforts will be aimed at cleaning the null values.

In [13]:
concat_df.dtypes

MS SubClass          int64
MS Zoning           object
Lot Frontage       float64
Lot Area             int64
Street              object
Alley               object
Lot Shape           object
Land Contour        object
Utilities           object
Lot Config          object
Land Slope          object
Neighborhood        object
Condition 1         object
Condition 2         object
Bldg Type           object
House Style         object
Overall Qual         int64
Overall Cond         int64
Roof Style          object
Roof Matl           object
Exterior 1st        object
Exterior 2nd        object
Mas Vnr Type        object
Mas Vnr Area       float64
Exter Qual          object
Exter Cond          object
Foundation          object
Bsmt Qual           object
Bsmt Cond           object
Bsmt Exposure       object
BsmtFin Type 1      object
BsmtFin SF 1       float64
BsmtFin Type 2      object
BsmtFin SF 2       float64
Bsmt Unf SF        float64
Total Bsmt SF      float64
Heating             object
H

In [14]:
# Investigate all null values
concat_df.isnull().sum()

# Create mask to find variables with null values
null_values = concat_df.isnull().sum()[concat_df.isnull().sum() > 0].sort_values(ascending=False)
null_values

Pool QC           2916
Misc Feature      2823
Alley             2731
Fence             2357
Fireplace Qu      1422
SalePrice          878
Lot Frontage       490
Garage Cond        159
Garage Qual        159
Garage Finish      159
Garage Yr Blt      159
Garage Type        157
Bsmt Exposure       83
BsmtFin Type 2      81
Bsmt Cond           80
Bsmt Qual           80
BsmtFin Type 1      80
Mas Vnr Area        23
Mas Vnr Type        23
Bsmt Half Bath       2
Bsmt Full Bath       2
BsmtFin SF 1         1
Electrical           1
Garage Cars          1
Garage Area          1
Total Bsmt SF        1
Bsmt Unf SF          1
BsmtFin SF 2         1
dtype: int64

In [15]:
# Create dictionary of null values
null_dict = {f'{col}': concat_df[col].isnull().sum() for col in null_values.index}

# Create percentages of null values in each variable
print('Null Values')
for key, value in null_dict.items():
    x = value / len(concat_df)
    print(f'{key}: {value}/{len(concat_df)} - {x*100:.2f}%')

Null Values
Pool QC: 2916/2929 - 99.56%
Misc Feature: 2823/2929 - 96.38%
Alley: 2731/2929 - 93.24%
Fence: 2357/2929 - 80.47%
Fireplace Qu: 1422/2929 - 48.55%
SalePrice: 878/2929 - 29.98%
Lot Frontage: 490/2929 - 16.73%
Garage Cond: 159/2929 - 5.43%
Garage Qual: 159/2929 - 5.43%
Garage Finish: 159/2929 - 5.43%
Garage Yr Blt: 159/2929 - 5.43%
Garage Type: 157/2929 - 5.36%
Bsmt Exposure: 83/2929 - 2.83%
BsmtFin Type 2: 81/2929 - 2.77%
Bsmt Cond: 80/2929 - 2.73%
Bsmt Qual: 80/2929 - 2.73%
BsmtFin Type 1: 80/2929 - 2.73%
Mas Vnr Area: 23/2929 - 0.79%
Mas Vnr Type: 23/2929 - 0.79%
Bsmt Half Bath: 2/2929 - 0.07%
Bsmt Full Bath: 2/2929 - 0.07%
BsmtFin SF 1: 1/2929 - 0.03%
Electrical: 1/2929 - 0.03%
Garage Cars: 1/2929 - 0.03%
Garage Area: 1/2929 - 0.03%
Total Bsmt SF: 1/2929 - 0.03%
Bsmt Unf SF: 1/2929 - 0.03%
BsmtFin SF 2: 1/2929 - 0.03%


In [16]:
# Create list of columns with zero values
zero_values = concat_df.isin([0]).sum().sort_values(ascending=False)
zero_values = zero_values[zero_values > 0]

# Create dictionary of zero values
zero_dict = {f'{col}': concat_df[col].isin([0]).sum() for col in zero_values.index}

# Find exact percentages of zero values
print('Zero Values')
for key, value in zero_dict.items():
    x = value / len(concat_df)
    print(f'{key}: {value}/{len(concat_df)} - {x*100:.2f}%')

Zero Values
Pool Area: 2916/2929 - 99.56%
3Ssn Porch: 2892/2929 - 98.74%
Low Qual Fin SF: 2889/2929 - 98.63%
Misc Val: 2826/2929 - 96.48%
Bsmt Half Bath: 2752/2929 - 93.96%
Screen Porch: 2673/2929 - 91.26%
BsmtFin SF 2: 2577/2929 - 87.98%
Enclosed Porch: 2470/2929 - 84.33%
Half Bath: 1843/2929 - 62.92%
Mas Vnr Area: 1748/2929 - 59.68%
Bsmt Full Bath: 1707/2929 - 58.28%
2nd Flr SF: 1678/2929 - 57.29%
Wood Deck SF: 1526/2929 - 52.10%
Fireplaces: 1422/2929 - 48.55%
Open Porch SF: 1300/2929 - 44.38%
BsmtFin SF 1: 930/2929 - 31.75%
Bsmt Unf SF: 244/2929 - 8.33%
Garage Cars: 157/2929 - 5.36%
Garage Area: 157/2929 - 5.36%
Age: 115/2929 - 3.93%
Total Bsmt SF: 79/2929 - 2.70%
Full Bath: 12/2929 - 0.41%
Bedroom AbvGr: 8/2929 - 0.27%
Kitchen AbvGr: 3/2929 - 0.10%


#### Top five null values handled by deductive imputation 

|Feature|Type|Explanation for NaN|Count of NaN|Deductive imputation basis|
|---|---|---|---|---|
|**Pool QC**|*Ordinal*|NA means no pool|2916|Matched to 2916 zero value in Pool Area
|**Misc Feature**|*Nominal*|NA means none|2823|Matched to 2826 zero value in Misc Val
|**Alley**|*Nominal*|NA means no alley access|2731|Only 7.33 miles of alley in large city [source](https://www.cityofames.org/government/departments-divisions-i-z/public-works/alley-maintenance)
|**Fence**|*Nominal*|NA means no fence|2357|Ames is urban area, likely that fencing not needed
|**Fireplace Qu**|*Nominal*|NA means no fence|1422|Matched wth 1422 zero value in Fireplaces

In [17]:
# For top 5 null value, I have filled the NaN with None, and converted them into to numbers, except for quality related features
# like Pool QC and Fireplace Qu
concat_df['Pool QC'] = concat_df['Pool QC'].fillna('None')
concat_df = pd.get_dummies(concat_df, columns=['Pool QC']).drop('Pool QC_None', axis=1)
concat_df = pd.get_dummies(concat_df, columns=['Misc Feature']).drop('Misc Feature_Othr', axis=1)
concat_df['Fence'] = concat_df['Fence'].fillna('None')
concat_df['Fence'] = concat_df['Fence'].map({'GdPrv':3,'MnPrv':2,'GdWo':1,'MnWw':1,'None':0})
concat_df['Alley'] = concat_df['Alley'].fillna('None')
concat_df = pd.get_dummies(concat_df, columns=['Alley']).drop('Alley_None', axis=1)
concat_df['Fireplace Qu'] = concat_df['Fireplace Qu'].fillna('None')
concat_df = pd.get_dummies(concat_df, columns=['Fireplace Qu']).drop('Fireplace Qu_None', axis=1)

#### Lot Frontage null value handling

Based on the percentages, there are 16.73% of null values for Lot Frontage. This is reasonable, as the property could be located in a high rise building, and thus is within expectation to have no Lot Frontage. As such, I will deduce that they should be zero.

I have done my external research, and found out that large apartment complexes or high rise apartments are the single most common housing type in Ames, accounting for 40.75% of the city's housing units. So it should be no surprise that so many properties have no Lot Frontage. 

Source: https://www.neighborhoodscout.com/ia/ames/real-estate#description

Only train data will be used, so as to prevent data leakage.

In [18]:
#Replace NaN with 0 for lot configuration
concat_df['Lot Frontage'] = concat_df['Lot Frontage'].fillna(0)

#### Electrical null value handling

In [19]:
# Find out which is the most frequent type for electrical - use train alone, to prevent data leakage
train['Electrical'].mode()

0    SBrkr
dtype: object

In [20]:
# Fill NaN with mode
concat_df['Electrical'] = concat_df['Electrical'].fillna('SBrkr')

#### Garage related null values

Noted the following are Garage related variables

1) Garage Cond

2) Garage Qual

3) Garage Finish

4) Garage Yr Blt

5) Garage Type

6) Garage Cars

7) Garage Area

The way that I will be performing cleaning of null values is further by deductive imputation. I manage to match 157 null values with zero values in Garage Area and Garage Cars. The last null value can be matched to the null value in Garage Area and Garage Cars which should have been 0. Thus I will be replacing all categorical null values to None and numerical null values to 0.

In [21]:
#Create list for garage categorical variables
garage_cat = ['Garage Type','Garage Finish','Garage Qual','Garage Cond']

#Create list for garage numerical variables
garage_num = ['Garage Yr Blt','Garage Cars', 'Garage Area']

In [22]:
# Replacing all garage categorical with None
for i in garage_cat:
    concat_df[i] = concat_df[i].replace(np.nan, 'None')

In [23]:
# Set all numerical to 0
for i in garage_num:
    concat_df[i] = concat_df[i].replace(np.nan, 0)

In [24]:
# Run getdummies for Garage type and drop None dummy column 
concat_df = pd.get_dummies(concat_df, columns=['Garage Type']).drop('Garage Type_None', axis=1)

#### Basement related null values

Noted the following are Basement related variables
1) Bsmt Exposure

2) BsmtFin Type 2

3) Bsmt Qual 

4) Bsmt Cond 

5) BsmtFin Type 1

6) Bsmt Full Bath 

7) Bsmt Half Bath

8) BsmtFin SF 1

9) BsmtFin SF 2

10) Bsmt Unf SF

11) Total Bsmt SF

By deductive imputation, I manage to match 79 null values with zero values in Total Bsmt SF. The last null value can be matched to the null value in Garage Area and Garage Cars which should have been 0. Thus I will be replacing all categorical null values to None and numerical null values to 0.

In [25]:
#Create list for basement categorical variables
bsmt_cat = ['Bsmt Exposure','BsmtFin Type 2','Bsmt Qual','Bsmt Cond', 'BsmtFin Type 1']

#Create list for basement numerical variables
bsmt_num = ['Bsmt Full Bath', 'Bsmt Half Bath', 'BsmtFin SF 1','BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF']

In [26]:
# Set basement null value numerical variables to 0
for i in bsmt_num:
    concat_df[i] = concat_df[i].replace(np.nan, 0)

In [27]:
# Set basement null value categorical variables to None
for i in bsmt_cat:
    concat_df[i] = concat_df[i].replace(np.nan, 'None')

In [28]:
# Make basement categorical variables into numbers other than Bsmt Cond and Qual
concat_df = pd.get_dummies(concat_df, columns=['Bsmt Cond']).drop('Bsmt Cond_None', axis=1)
concat_df['Bsmt Exposure'] = concat_df['Bsmt Exposure'].map({'Gd':3,'Av':2,'Mn':1,'No':0,'None':0})
concat_df = pd.get_dummies(concat_df, columns=['Bsmt Qual']).drop('Bsmt Qual_None', axis=1)
concat_df['BsmtFin Type 1'] = concat_df['BsmtFin Type 1'].map({'GLQ':3,'ALQ':2,'BLQ':1,'Rec':1,'LwQ':0.5,'Unf':0,'None':-0.5})
concat_df['BsmtFin Type 2'] = concat_df['BsmtFin Type 2'].map({'GLQ':3,'ALQ':2,'BLQ':1,'Rec':1,'LwQ':0.5,'Unf':0,'None':-0.5})

#### Masonry related null values

As there is not much data from zero values that allows me to do deductive imputation, I am opting to make an educated guess that masonry related null values are None or 0 values. Afterall there are not many null values (only 1.07%), so from the perspective of efficiency, it should not affect the training test set by much.

In [29]:
# Replace numerical masonry related variable to 0  
concat_df['Mas Vnr Area'] = concat_df['Mas Vnr Area'].replace(np.nan, 0)

# Replace categorical masonry related variable to None 
concat_df['Mas Vnr Type'] = concat_df['Mas Vnr Type'].replace(np.nan, None)

In [30]:
# Run getdummies for Mas Vnr type and drop None dummy column 
concat_df = pd.get_dummies(data = concat_df, columns=['Mas Vnr Type']).drop('Mas Vnr Type_None', axis=1)

At this point, we are left with 27 categorical variables to clean up further into numeric features. Of these remaining features, 10 are ordinal features and 17 are nominal features. 

In [31]:
# Changing remaining 10 ordinal features to numerical, other than Functional, Kitchen Qual and Exter Qual,
# these are important features which I will make get dummies to make them more accurate in scale and weightage
concat_df['Lot Shape'] = concat_df['Lot Shape'].map({'IR1':1,'IR2':1,'IR3':1,'Reg':0})
concat_df['Land Slope'] = concat_df['Land Slope'].map({'Gtl':0,'Mod':1,'Sev':1})
concat_df['Utilities'] = concat_df['Utilities'].map({'AllPub':0,'NoSewr':-1,'NoSeWa':-1})
concat_df['Heating QC'] = concat_df['Heating QC'].map({'Ex':4,'Gd':3,'TA':2,'Fa':1,'Po':0})
concat_df['Electrical'] = concat_df['Electrical'].map({'SBrkr':1,'FuseA':0.5,'FuseF':0,'FuseP':0,'Mix':0})
concat_df['Exter Cond'] = concat_df['Exter Cond'].map({'Ex':2,'Gd':1,'TA':0,'Fa':-1,'Po':-2})
concat_df['Paved Drive'] = concat_df['Paved Drive'].map({'Y':1,'P':0.5,'N':0})
concat_df = pd.get_dummies(data = concat_df, columns=['Functional']).drop('Functional_Typ', axis=1)
concat_df = pd.get_dummies(data = concat_df, columns=['Kitchen Qual']).drop('Kitchen Qual_Fa', axis=1)
concat_df = pd.get_dummies(data = concat_df, columns=['Exter Qual']).drop('Exter Qual_Fa', axis=1)

Weightages are given to each variable based on preliminary exploratory data analysis.For variables that do not seem to have any relationship, lesser weight is given to each category. For those identified in boxplot, higher weightages are given. For Functional, which is more complex, I will allow get dummies to deal with it, and my machine learning model in feature engineering to deal with it.

In [32]:
# Collect remaining features
cat_features = [col for col in concat_df.select_dtypes(include=['category', 'object']).columns]

In [33]:
concat_df = pd.get_dummies(data = concat_df, columns=cat_features)

In [34]:
# Numerical to categorical
concat_df = pd.get_dummies(data = concat_df, columns=['MS SubClass']).drop('MS SubClass_20', axis=1)

In [35]:
# Numerical to categorical
concat_df = pd.get_dummies(data = concat_df, columns=['Mo Sold']).drop('Mo Sold_1', axis=1)

In [36]:
concat_df.shape

(2929, 282)

### Outliers 
A few outliers were also visibly noted in the following features and will be drilled down further to understand:

1) BsmtFin SF 1

2) Gr Liv Area

3) Garage Yr Blt

The general approach used is to understand if there are clearly any mistakes in the data. If no mistake can be deduced, the data will be preserved as I will be using it for prediction purposes.

In [37]:
#2) 2 visible outliers for BsmtFin SF 1
train[train['BsmtFin SF 1']>3000]

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
960   1499  908154235           60        RL         313.0     63887   Pave   
1885  2181  908154195           20        RL         128.0     39290   Pave   

     Alley Lot Shape Land Contour Utilities Lot Config Land Slope  \
960    NaN       IR3          Bnk    AllPub     Corner        Gtl   
1885   NaN       IR1          Bnk    AllPub     Inside        Gtl   

     Neighborhood Condition 1 Condition 2 Bldg Type House Style  Overall Qual  \
960       Edwards       Feedr        Norm      1Fam      2Story            10   
1885      Edwards        Norm        Norm      1Fam      1Story            10   

      Overall Cond  Year Built  Year Remod/Add Roof Style Roof Matl  \
960              5        2008            2008        Hip   ClyTile   
1885             5        2008            2009        Hip   CompShg   

     Exterior 1st Exterior 2nd Mas Vnr Type  Mas Vnr Area Exter Qual  \
960        Stucco       Stucco        Stone         796.0         Ex   
1885      CemntBd      CmentBd        Stone        1224.0         Ex   

     Exter Cond Foundation Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  \
960          TA      PConc        Ex        TA            Gd            GLQ   
1885         TA      PConc        Ex        TA            Gd            GLQ   

      BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  \
960         5644.0            Unf           0.0        466.0         6110.0   
1885        4010.0            Unf           0.0       1085.0         5095.0   

     Heating Heating QC Central Air Electrical  1st Flr SF  2nd Flr SF  \
960     GasA         Ex           Y      SBrkr        4692         950   
1885    GasA         Ex           Y      SBrkr        5095           0   

      Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  Bsmt Half Bath  Full Bath  \
960                 0         5642             2.0             0.0          2   
1885                0         5095             1.0             1.0          2   

      Half Bath  Bedroom AbvGr  Kitchen AbvGr Kitchen Qual  TotRms AbvGrd  \
960           1              3              1           Ex             12   
1885          1              2              1           Ex             15   

     Functional  Fireplaces Fireplace Qu Garage Type  Garage Yr Blt  \
960         Typ           3           Gd      Attchd         2008.0   
1885        Typ           2           Gd      Attchd         2008.0   

     Garage Finish  Garage Cars  Garage Area Garage Qual Garage Cond  \
960            Fin          2.0       1418.0          TA          TA   
1885           Fin          3.0       1154.0          TA          TA   

     Paved Drive  Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  \
960            Y           214            292               0           0   
1885           Y           546            484               0           0   

      Screen Porch  Pool Area Pool QC Fence Misc Feature  Misc Val  Mo Sold  \
960              0        480      Gd   NaN          NaN         0        1   
1885             0          0     NaN   NaN         Elev     17000       10   

      Yr Sold Sale Type  SalePrice  
960      2008       New     160000  
1885     2007       New     183850

In [38]:
#3) 2 visible outliers for Gr Liv Area
train[train['Gr Liv Area']>4000]

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
960   1499  908154235           60        RL         313.0     63887   Pave   
1885  2181  908154195           20        RL         128.0     39290   Pave   

     Alley Lot Shape Land Contour Utilities Lot Config Land Slope  \
960    NaN       IR3          Bnk    AllPub     Corner        Gtl   
1885   NaN       IR1          Bnk    AllPub     Inside        Gtl   

     Neighborhood Condition 1 Condition 2 Bldg Type House Style  Overall Qual  \
960       Edwards       Feedr        Norm      1Fam      2Story            10   
1885      Edwards        Norm        Norm      1Fam      1Story            10   

      Overall Cond  Year Built  Year Remod/Add Roof Style Roof Matl  \
960              5        2008            2008        Hip   ClyTile   
1885             5        2008            2009        Hip   CompShg   

     Exterior 1st Exterior 2nd Mas Vnr Type  Mas Vnr Area Exter Qual  \
960        Stucco       Stucco        Stone         796.0         Ex   
1885      CemntBd      CmentBd        Stone        1224.0         Ex   

     Exter Cond Foundation Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  \
960          TA      PConc        Ex        TA            Gd            GLQ   
1885         TA      PConc        Ex        TA            Gd            GLQ   

      BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  \
960         5644.0            Unf           0.0        466.0         6110.0   
1885        4010.0            Unf           0.0       1085.0         5095.0   

     Heating Heating QC Central Air Electrical  1st Flr SF  2nd Flr SF  \
960     GasA         Ex           Y      SBrkr        4692         950   
1885    GasA         Ex           Y      SBrkr        5095           0   

      Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  Bsmt Half Bath  Full Bath  \
960                 0         5642             2.0             0.0          2   
1885                0         5095             1.0             1.0          2   

      Half Bath  Bedroom AbvGr  Kitchen AbvGr Kitchen Qual  TotRms AbvGrd  \
960           1              3              1           Ex             12   
1885          1              2              1           Ex             15   

     Functional  Fireplaces Fireplace Qu Garage Type  Garage Yr Blt  \
960         Typ           3           Gd      Attchd         2008.0   
1885        Typ           2           Gd      Attchd         2008.0   

     Garage Finish  Garage Cars  Garage Area Garage Qual Garage Cond  \
960            Fin          2.0       1418.0          TA          TA   
1885           Fin          3.0       1154.0          TA          TA   

     Paved Drive  Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  \
960            Y           214            292               0           0   
1885           Y           546            484               0           0   

      Screen Porch  Pool Area Pool QC Fence Misc Feature  Misc Val  Mo Sold  \
960              0        480      Gd   NaN          NaN         0        1   
1885             0          0     NaN   NaN         Elev     17000       10   

      Yr Sold Sale Type  SalePrice  
960      2008       New     160000  
1885     2007       New     183850

The source document mentioned to remove houses that are more than 4,000 square feet, to remove the unusual transactions. ID 1499 and 2181 has been spotted as outliers for both BsmtFin SF 1 and Gr Liv Area. However, unlike the five observations that was mentioned by the source document, I have only managed to spot two. I will be removing the outliers to make the model more accurate.

In [39]:
concat_df.drop([960, 1885], inplace = True)

In [40]:
test[test['Gr Liv Area']>4000]

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
170  1761  528320050           60        RL         160.0     15623   Pave   
463  1768  528351010           60        RL         104.0     21535   Pave   

    Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
170   NaN       IR1          Lvl    AllPub     Corner        Gtl      NoRidge   
463   NaN       IR1          Lvl    AllPub     Corner        Gtl      NoRidge   

    Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
170        Norm        Norm      1Fam      2Story            10             5   
463        Norm        Norm      1Fam      2Story            10             6   

     Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st  \
170        1996            1996        Hip   CompShg      Wd Sdng   
463        1994            1995      Gable   WdShngl      HdBoard   

    Exterior 2nd Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation  \
170      ImStucc         None           0.0         Gd         TA      PConc   
463      HdBoard      BrkFace        1170.0         Ex         TA      PConc   

    Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
170        Ex        TA            Av            GLQ          2096   
463        Ex        TA            Gd            GLQ          1455   

    BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating  \
170            Unf             0          300           2396    GasA   
463            Unf             0          989           2444    GasA   

    Heating QC Central Air Electrical  1st Flr SF  2nd Flr SF  \
170         Ex           Y      SBrkr        2411        2065   
463         Ex           Y      SBrkr        2444        1872   

     Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  Bsmt Half Bath  Full Bath  \
170                0         4476               1               0          3   
463                0         4316               0               1          3   

     Half Bath  Bedroom AbvGr  Kitchen AbvGr Kitchen Qual  TotRms AbvGrd  \
170          1              4              1           Ex             10   
463          1              4              1           Ex             10   

    Functional  Fireplaces Fireplace Qu Garage Type  Garage Yr Blt  \
170        Typ           2           TA      Attchd         1996.0   
463        Typ           2           Ex      Attchd         1994.0   

    Garage Finish  Garage Cars  Garage Area Garage Qual Garage Cond  \
170           Fin            3          813          TA          TA   
463           Fin            3          832          TA          TA   

    Paved Drive  Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  \
170           Y           171             78               0           0   
463           Y           382             50               0           0   

     Screen Porch  Pool Area Pool QC  Fence Misc Feature  Misc Val  Mo Sold  \
170             0        555      Ex  MnPrv          NaN         0        7   
463             0          0     NaN    NaN          NaN         0        1   

     Yr Sold Sale Type  
170     2007       WD   
463     2007       WD

I have verified that there are another two observations that are more than 4,000 square feet in the test dataset. Test dataset will be left in tact as they will be required for Kaggle submission.

In [41]:
#4) 1 visible outlier for Garage Yr Blt in 2207
train[train['Garage Yr Blt']>2100]

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
1699  2261  916384070           20        RL          68.0      8298   Pave   

     Alley Lot Shape Land Contour Utilities Lot Config Land Slope  \
1699   NaN       IR1          HLS    AllPub     Inside        Gtl   

     Neighborhood Condition 1 Condition 2 Bldg Type House Style  Overall Qual  \
1699       Timber        Norm        Norm      1Fam      1Story             8   

      Overall Cond  Year Built  Year Remod/Add Roof Style Roof Matl  \
1699             5        2006            2007        Hip   CompShg   

     Exterior 1st Exterior 2nd Mas Vnr Type  Mas Vnr Area Exter Qual  \
1699      VinylSd      VinylSd          NaN           NaN         Gd   

     Exter Cond Foundation Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  \
1699         TA      PConc        Gd        TA            Av            GLQ   

      BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  \
1699         583.0            Unf           0.0        963.0         1546.0   

     Heating Heating QC Central Air Electrical  1st Flr SF  2nd Flr SF  \
1699    GasA         Ex           Y      SBrkr        1564           0   

      Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  Bsmt Half Bath  Full Bath  \
1699                0         1564             0.0             0.0          2   

      Half Bath  Bedroom AbvGr  Kitchen AbvGr Kitchen Qual  TotRms AbvGrd  \
1699          0              2              1           Ex              6   

     Functional  Fireplaces Fireplace Qu Garage Type  Garage Yr Blt  \
1699        Typ           1           Gd      Attchd         2207.0   

     Garage Finish  Garage Cars  Garage Area Garage Qual Garage Cond  \
1699           RFn          2.0        502.0          TA          TA   

     Paved Drive  Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  \
1699           Y           132              0               0           0   

      Screen Porch  Pool Area Pool QC Fence Misc Feature  Misc Val  Mo Sold  \
1699             0          0     NaN   NaN          NaN         0        9   

      Yr Sold Sale Type  SalePrice  
1699     2007       New     267300

This looked to me like a case of fat finger mistake, and I will be replacing it to 2007, which is the year that Year Remod/Add was done. The house was also first built in 2006, so it is reasonable for replacement value to be 2007.

In [42]:
concat_df['Garage Yr Blt'].replace({2207: 2007}, inplace=True)

In [43]:
concat_df = concat_df.reset_index(drop=True)

## Outlier from error analysis

In [48]:
train.iloc[125]

Id                      1183
PID                533350090
MS SubClass               60
MS Zoning                 RL
Lot Frontage             NaN
Lot Area               24572
Street                  Pave
Alley                    NaN
Lot Shape                IR1
Land Contour             Lvl
Utilities             AllPub
Lot Config           CulDSac
Land Slope               Gtl
Neighborhood         Veenker
Condition 1             Norm
Condition 2             Norm
Bldg Type               1Fam
House Style           2Story
Overall Qual               9
Overall Cond               3
Year Built              1977
Year Remod/Add          1977
Roof Style           Mansard
Roof Matl            CompShg
Exterior 1st         Wd Sdng
Exterior 2nd         Wd Sdng
Mas Vnr Type         BrkFace
Mas Vnr Area          1050.0
Exter Qual                Gd
Exter Cond                Gd
Foundation            CBlock
Bsmt Qual                 Gd
Bsmt Cond                 TA
Bsmt Exposure             No
BsmtFin Type 1

In [44]:
concat_df.iloc[125]

Lot Frontage                 0.0
Lot Area                 24572.0
Lot Shape                    1.0
Utilities                    0.0
Land Slope                   0.0
Overall Qual                 9.0
Overall Cond                 3.0
Mas Vnr Area              1050.0
Exter Cond                   1.0
Bsmt Exposure                0.0
BsmtFin Type 1               3.0
BsmtFin SF 1               410.0
BsmtFin Type 2               0.0
BsmtFin SF 2                 0.0
Bsmt Unf SF                584.0
Total Bsmt SF              994.0
Heating QC                   2.0
Electrical                   1.0
1st Flr SF                1599.0
2nd Flr SF                1345.0
Low Qual Fin SF              0.0
Gr Liv Area               2944.0
Bsmt Full Bath               0.0
Bsmt Half Bath               0.0
Full Bath                    2.0
Half Bath                    2.0
Bedroom AbvGr                3.0
Kitchen AbvGr                1.0
TotRms AbvGrd                9.0
Fireplaces                   1.0
Garage Yr 

In [45]:
concat_df.drop(125, inplace = True)

These is an outlier that are spotted when residual plot was performed with final model. Under the assumption that these properties were not sold at arms length, causing it to be severely undervalued, I will be removing them from the dataset. I have verified that the resulting RMSE and residual plot becomes more reasonable after their removal.

## Saving of cleaned CSV files

In [46]:
concat_df.shape

(2926, 282)

In [47]:
# Save cleaned test data
concat_df.to_csv('./datasets/concat_df_cleaned.csv', index=False)